### PR TITLE
[SPARK-14929][SQL] Disable vectorized map for wide schemas & high-precision decimals

### PR DESCRIPTION
**Pull request by sameeragarwal**
## What changes were proposed in this pull request?

While the vectorized hash map in `TungstenAggregate` is currently supported for all primitive data types during partial aggregation, this patch only enables the hash map for a subset of cases that've been verified to show performance improvements on our benchmarks subject to an internal conf that sets an upper limit on the maximum length of the aggregate key/value schema. This list of supported use-cases should be expanded over time.
## How was this patch tested?

This is no new change in functionality so existing tests should suffice. Performance tests were done on TPCDS benchmarks.

### CODE REVIEW

**Comment by sameeragarwal** [comment]
cc @davies

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregate.scala [hunk 1]
```diff
@@ -447,16 +447,31 @@ case class TungstenAggregate(
...
+   * subject to an internal conf that sets an upper limit on the maximum length of the aggregate
+   * key/value schema.
+   *
+   * This list of supported use-cases should be expanded over time.
+   */
+  private def enableVectorizedHashMap(ctx: CodegenContext): Boolean = {
+    val isSupported =
+      (groupingKeySchema ++ bufferSchema).forall(f => ctx.isPrimitiveType(f.dataType) ||
+        f.dataType.isInstanceOf[DecimalType] || f.dataType.isInstanceOf[StringType]) &&
+        bufferSchema.forall(!_.dataType.isInstanceOf[StringType]) && bufferSchema.nonEmpty &&
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregate.scala]: bufferSchema can't have StringType in TungstenAggregate

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregate.scala [hunk 2]
```diff
@@ -447,16 +447,31 @@ case class TungstenAggregate(
...
+        modes.forall(mode => mode == Partial || mode == PartialMerge)
+
+    isSupported && bufferSchema.map(_.dataType).filter(_.isInstanceOf[DecimalType])
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregate.scala]: Add a comment to say why ByteArrayDecimalType is not supported?

**Comment by davies** [comment]
LGTM
Merging this into master, thanks!

**Pull request closed without merging**